src/main/java/com/czertainly/core/service/cmp/impl/CmpServiceImpl.java

```diff
@@ -0,0 +1,344 @@
+                    throw new CmpProcessingException(PKIFailureInfo.badRequest,
+                            ImplFailureInfo.CMPVALMSG201);
+            }
+
+            if(pkiResponse == null) {
+                throw new CmpProcessingException(
+                        PKIFailureInfo.systemFailure,
+                        String.format(" %s | general problem while handling PKIMessage", logPrefix));
+            }
+            LOG.info("{} | response processed: {}",
+                    PkiMessageDumper.logPrefix(pkiResponse, profileName),
+                    PkiMessageDumper.dumpPkiMessage(verbose, pkiResponse));
+
+            headerValidator.validate(pkiResponse, config3gppProfile);
+            bodyValidator.validate(pkiResponse, config3gppProfile);
+            protectionValidator.validateOut(pkiResponse, config3gppProfile);
+
+            return buildOk(pkiResponse);
+        } catch (CmpBaseException e) {
+            PKIMessage pkiResponse = PkiMessageError.unprotectedMessage(pkiRequest.getHeader(), e.toPKIBody());
+            if(verbose) {
+                LOG.error("{} | processing failed: \n\n response:\n {}", logPrefix,
+                        PkiMessageDumper.dumpPkiMessage(pkiResponse), e);
+            } else {
+                LOG.error("{} | processing failed: \n\nrequest:\n {}\n response:\n {}", logPrefix,
+                        requestAsString, PkiMessageDumper.dumpPkiMessage(pkiResponse), e);
+            }
+            return buildBadRequest(pkiResponse);
+        } catch (IOException e) {
+            PKIMessage pkiResponse = PkiMessageError.unprotectedMessage(
+                    pkiRequest.getHeader(),
+                    PKIFailureInfo.badDataFormat,
+                    ImplFailureInfo.CMPSRV101);
+            if(verbose) {
+                LOG.error("{} | parsing failed: \n\n response:\n {}", logPrefix,
+                        PkiMessageDumper.dumpPkiMessage(pkiResponse), e);
+            } else {
+                LOG.error("{} | parsing failed: \n\nrequest:\n {}\n response:\n {}", logPrefix,
+                        requestAsString, PkiMessageDumper.dumpPkiMessage(pkiResponse), e);
+            }
+            return buildBadRequest(pkiResponse);
+        } catch (Exception e) {
+            PKIMessage pkiResponse = PkiMessageError.unprotectedMessage(pkiRequest.getHeader(), e);
+            if(verbose) {
+                LOG.error("{} | handling failed: \n\n response:\n {}", logPrefix,
+                        PkiMessageDumper.dumpPkiMessage(pkiResponse), e);
+            } else {
+                LOG.error("{} | handling failed: \n\nrequest:\n {}\n response:\n {}", logPrefix,
+                        requestAsString, PkiMessageDumper.dumpPkiMessage(pkiResponse), e);
+            }
+            return buildBadRequest(pkiResponse);
+        }
+    }
+
+    private ResponseEntity buildBadRequest(PKIMessage pkiMessage) {

```

Comment:
Raw use of parameterized class 'ResponseEntity'

Reply:
I do not understand this topic/point, please a little detail - what is wrong using ResponseEntity this way?